File: March2022/Java/CopyLLwithRandomPointers.java
// Author: Shobhit Behl (LC: shobhitbruh)
class Solution {
    public Node copyRandomList(Node head) {
        Node p=head;
        HashMap<Node,Node> hm=new HashMap<>();
        
        while(p!=null){
            Node x=new Node(p.val);
            hm.put(p,x);
            p=p.next;
        }
        
        p=head;
        while(p!=null){
            Node x=hm.get(p);
            if(p.next!=null){
                x.next=hm.get(p.next);
            }
            if(p.random!=null){
                x.random=hm.get(p.random);
            }
            p=p.next;
        }
        return hm.get(head);
        
    }
}

File: March2022/Kotlin/CopyLLwithRandomPointers.kt
// Author: Shobhit Behl (LC: shobhitbruh)
fun copyRandomList(head: Node?): Node? {
        var p = head
        val hm = HashMap<Node?, Node>()
        while (p != null) {
            val x: Node = Node(p.`val`)
            hm[p] = x
            p = p.next
        }
        p = head
        while (p != null) {
            val x = hm[p]
            if (p.next != null) {
                x!!.next = hm[p.next]
            }
            if (p.random != null) {
                x!!.random = hm[p.random]
            }
            p = p.next
        }
        return hm[head]
    }